spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan resolveOperators {
+      case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
+          if matchedActions.isEmpty =>
+
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("merge", newWriteInfo(target.schema))
+        val targetTableScan = buildSimpleScanPlan(target.table, target.output, mergeBuilder, cond)
+
+        // when there are no matched actions, use a left anti join to remove any matching rows and rewrite to use
+        // append instead of replace. only unmatched source rows are passed to the merge and actions are all inserts.
+        val joinPlan = Join(source, targetTableScan, LeftAnti, Some(cond), JoinHint.NONE)
+
+        val mergeParams = MergeIntoParams(
```

Comment:
Is there a particular reason to use `MergeInto` node here? I think a simple projection will be enough and it will support codegen.

Reply:
You mean we should rewrite this to `CASE ... WHEN`?

Reply:
My assumption was that there may be only one WHEN NOT MATCHED clause which is not true starting from Spark 3.1.0. If we have only one, though, we can just use a projection. I guess CASE ... WHEN approach can work. Let's do it later.

Reply:
To follow up on this, it is possible to have multiple `NOT MATCHED` clauses. So it is a good idea to use the merge here. We can support another case later.

Reply:
+1, let me resolve this for now.